graphqlapi/src/main/java/io/stargate/graphql/schema/graphqlfirst/processor/IncrementModelBuilder.java

```diff
@@ -0,0 +1,146 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.stargate.graphql.schema.graphqlfirst.processor;
+
+import graphql.Scalars;
+import graphql.language.Directive;
+import graphql.language.InputValueDefinition;
+import graphql.language.ListType;
+import graphql.language.Type;
+import graphql.language.TypeName;
+import graphql.schema.GraphQLScalarType;
+import io.stargate.graphql.schema.graphqlfirst.util.TypeHelper;
+import io.stargate.graphql.schema.scalars.CqlScalar;
+import java.util.*;
+import java.util.stream.Collectors;
+
+public class IncrementModelBuilder extends ModelBuilderBase<IncrementModel> {
+
+  public static final String COUNTER_TYPE_NAME = "counter";
+  private final EntityModel entity;
+  private final FieldModel field;
+  private final Map<String, EntityModel> entities;
+  private final ProcessingContext context;
+
+  private static final boolean PREPEND_DEFAULT = false;
+  private final String operationName;
+  private final InputValueDefinition argument;
+
+  public IncrementModelBuilder(
+      InputValueDefinition argument,
+      String operationName,
+      EntityModel entity,
+      FieldModel field,
+      Map<String, EntityModel> entities,
+      ProcessingContext context) {
+    super(context, argument.getSourceLocation());
+    this.argument = argument;
+    this.operationName = operationName;
+    this.entity = entity;
+    this.field = field;
+    this.entities = entities;
+    this.context = context;
+  }
+
+  @Override
+  IncrementModel build() throws SkipException {
+
+    Optional<Directive> directive = DirectiveHelper.getDirective(CqlDirectives.INCREMENT, argument);
+
+    boolean prepend =
+        directive
+            .flatMap(d -> DirectiveHelper.getBooleanArgument(d, "prepend", context))
+            .orElse(PREPEND_DEFAULT);
+
+    validate(field, prepend);
+    return new IncrementModel(field, prepend, argument.getName());
+  }
+
+  protected void validate(FieldModel field, boolean prepend) throws SkipException {
+    if (field.isPartitionKey() || field.isClusteringColumn()) {
+      invalidMapping(
+          "Operation %s: directive %s is not supported for partition/clustering key field %s.",
+          operationName, CqlDirectives.INCREMENT, field.getGraphqlName());
+      throw SkipException.INSTANCE;
+    } else {
+      checkValidForRegularColumn(field, prepend);
+    }
+  }
+
+  private void checkValidForRegularColumn(FieldModel field, boolean prepend) throws SkipException {
+
+    Type<?> fieldInputType =
+        toInput(field.getGraphqlType(), argument, entity, field, entities, operationName);
+    // if it is non-list type
+    if (fieldInputType instanceof TypeName) {
+      validateCounter((TypeName) fieldInputType);
+      validatePrepend(field, prepend);
+    } else {
+      // otherwise it must be a list
+      checkArgumentIsListOf(argument, entity, entities, operationName, field);
```

Comment:
This not the right check.

What that method does is check that the argument _is a list of_ the field type (e.g. field = `Int` => argument = `[Int]`).

What we want is to check that both the argument and the field are lists of the same element type. For example:
```graphql
    type List {
      id: ID
      l: [Int]
    }
    type Mutation {
      updateList(
        id: ID
        lSuffix: [Int] @cql_increment(field: "l")
      ): Boolean @cql_update(targetEntity: "List")
    }
```

PS: I tried to deploy the above and to my surprise it worked. Turns out that there is a bug in `checkArgumentIsListOf`. I'll open a separate ticket to fix it.

Reply:
fixed in:
https://github.com/stargate/stargate/pull/1015

Reply:
Even after fixing the method, we still have to change the call here, as I said it's not the right check. Also addressed in 3e2b33f0.